File: src/main/java/com/ajax/springcourse/car/observable/GrpcCarStreamObservable.java
package com.ajax.springcourse.car.observable;

import com.ajax.springcourse.car.model.dto.CarReadDto;
import com.ajax.springcourse.grpc.Car;
import io.grpc.stub.StreamObserver;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
@Slf4j
public class GrpcCarStreamObservable {
    private final List<StreamObserver<Car.CarReadDto>> observers = new ArrayList<>();

    public void addObserver(StreamObserver<Car.CarReadDto> observer){
        log.info("Adding observer "+observer.toString());
        observers.add(observer);
    }

    public void removeObserver(StreamObserver<Car.CarReadDto> observer){
        log.info("Removing observer "+observer.toString());
        observers.remove(observer);
    }

    public void newCarReadDto(Car.CarReadDto carReadDto){
        log.info("Pushing new car with id="+carReadDto.getId());
        observers.forEach(observer -> observer.onNext(carReadDto));
    }
}

File: src/main/java/com/ajax/springcourse/car/observable/GrpcCarStreamObservable.kt
package com.ajax.springcourse.car.observable

import com.ajax.springcourse.grpc.Car
import io.grpc.stub.StreamObserver
import io.nats.client.Connection
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.util.SerializationUtils
import javax.annotation.PostConstruct

@Component
class GrpcCarStreamObservable @Autowired constructor(val connection: Connection) {

    companion object {
        const val SUBJECT = "carReadDto.create"
    }

    private val logger = LoggerFactory.getLogger(this::class.java)

    private val observers: MutableList<StreamObserver<Car.CarReadDto>> = ArrayList()

    @PostConstruct
    fun init() {
        val dispatcher = connection.createDispatcher { msg ->
            val carReadDto = SerializationUtils.deserialize(msg.data) as Car.CarReadDto
            logger.info("Got car with id=${carReadDto.id}")
            observers.forEach { it.onNext(carReadDto) }
        }
        dispatcher.subscribe(SUBJECT)
    }

    fun addObserver(observer: StreamObserver<Car.CarReadDto>) {
        observers.add(observer)
    }

    fun removeObserver(observer: StreamObserver<Car.CarReadDto>) {
        observers.remove(observer)
    }

    fun newCarReadDto(carReadDto: Car.CarReadDto) {
        logger.info("Sending car with id=${carReadDto.id}")
        connection.publish(SUBJECT, SerializationUtils.serialize(carReadDto))
    }
}